### ISSUE TITLE
gdb-dashboard unresponsive upon receiving SIGSEGV on binary being debugged

### ISSUE THREAD

**Issue by Rubonnek** [cyrus-and/gdb-dashboard#210]
Using the [Godot Game Engine](https://github.com/godotengine/godot) at commit 4abebfa476c54552f55b03bb43d72bd29a7a26b1

```shell
git clone https://github.com/godotengine/godot
cd godot
git checkout 4abebfa476c54552f55b03bb43d72bd29a7a26b1
scons platform=x11 CCFLAGS='-D_FORTIFY_SOURCE=2' CXX=g++ target=release_debug debug_release=yes debug_symbols=full tools=yes bits=64 -j$(nproc)
```

And opening the following project which causes a crash: [xml-multiline-comment.zip](https://github.com/cyrus-and/gdb-dashboard/files/4927069/xml-multiline-comment.zip)

```shell
unzip xml-multiline-comment.zip
cd xml-multiline-comment
gdb /path/to/godot.x11.opt.tools.64 # it's a stack smash -- may have to run several time to trigger the crash
```

Makes `gdb-dashboard` unresponsive for a good while. Doesn't seem related to #208, but could be.

**Comment by cyrus-and** [comment]
Thanks for the super detailed instructions, I really appreciate your time.

I tried in a Docker container using Xvfb and I cannot make it crash, I just get the following then the engine waits for something I guess.

```
user@6c6de16254b0:~/xml-multiline-comment$ /home/user/godot/bin/godot.x11.opt.tools.64
Godot Engine v3.2.3.beta.custom_build.4abebfa47 - https://godotengine.org
OpenGL ES 3.0 Renderer: llvmpipe (LLVM 7.0, 256 bits)
ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4568:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5047:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2565:(snd_pcm_open_noupdate) Unknown PCM default
ERROR: init_device: Condition "status < 0" is true. Returned: ERR_CANT_OPEN
   At: drivers/alsa/audio_driver_alsa.cpp:83.
WARNING: initialize: All audio drivers failed, falling back to the dummy driver.
   At: servers/audio_server.cpp:215.
```

I will try on a regular Linux later on. In the meanwhile...

1. Please tell me where the program stops after the SIGSEGV, the precise location where the dashboard becomes irresponsive so I can try to place a breakpoint there (or nearby, if this issue is really related to #208 it should be enough...).

2. Also, once the dashboard finally loads, what if you start stepping? Does every time it becomes unresponsive?

3. Finally, try to reproduce is without the Assembly module then without the Source module. Does the issue still happen in these cases? Run one of the following before starting the program.

    ```
    dashboard -layout breakpoints expressions history memory registers source stack threads variables
    dashboard -layout assembly breakpoints expressions history memory registers stack threads variables
    ```

**Comment by Rubonnek** [comment]
Here's where it stops:
```
Thread 1 "godot" received signal SIGSEGV, Segmentation fault.
XMLParser::_parse_comment (this=0x76444c0) at core/io/xml_parser.cpp:203
203                     if (*P == '>')
```

Yes it becomes unresponsive when I try to step.

Runs fine when disabling the Assembly module. Still slow when only disabling the Source module.

Looks like this is a duplicate of #208 after all.

**Comment by cyrus-and** [comment]
Yes there are some cases in which GDB fails to properly locate the function bounds or the function is actually *huge* so the disassembling + highlighting takes a lot of time, I should cache the result or avoid processing *off-screen* instructions. I remember there were some non-trivial hinderances... but I never came up with a reliable way to reproduce it.

I'll look into it and thanks once again for your support.

P.S. it might help to disable the syntax highlighting in the meantime:

```
dashboard -style syntax_highlighting None
```

**Comment by cyrus-and** [comment]
Alright, I'm finally able to reproduce this:

```
>>> break core/io/xml_parser.cpp:203
>>> info breakpoints
Num     Type           Disp Enb Address            What
1       breakpoint     keep y   0x0000000002bb88fa in XMLParser::_parse_comment() at core/io/xml_parser.cpp:203
>>> starti
>>> set $pc=0x0000000002bb88fa
>>> dashboard
```

**Comment by cyrus-and** [comment]
And here's the cause:

```
>>> disassemble
Dump of assembler code for function XMLParser::_parse_comment():
Address range 0x2bb88b0 to 0x2bb8999:
   0x0000000002bb88b0 <+0>:	push   %r13
   0x0000000002bb88b2 <+2>:	push   %r12
   0x0000000002bb88b4 <+4>:	push   %rbp
   ...
   0x0000000002bb898c <+220>:	jmpq   0x754e41 <XMLParser::_parse_comment()+4256810385>
   0x0000000002bb8991 <+225>:	mov    %rax,%rbx
   0x0000000002bb8994 <+228>:	jmpq   0x754e34 <XMLParser::_parse_comment()+4256810372>
Address range 0x754e34 to 0x754e55:
   0x0000000000754e34 <+-38156924>:	mov    0x8(%rsp),%rsi
   0x0000000000754e39 <+-38156919>:	mov    %r12,%rdi
   0x0000000000754e3c <+-38156916>:	callq  0x2bb78b0 <CowData<wchar_t>::_unref(void*)>
   0x0000000000754e41 <+-38156911>:	mov    (%rsp),%rsi
   0x0000000000754e45 <+-38156907>:	mov    %rbp,%rdi
   0x0000000000754e48 <+-38156904>:	callq  0x2bb78b0 <CowData<wchar_t>::_unref(void*)>
   0x0000000000754e4d <+-38156899>:	mov    %rbx,%rdi
   0x0000000000754e50 <+-38156896>:	callq  0x4318c0 <_Unwind_Resume@plt>
End of assembler dump.
```

That function spans two non-contiguous blocks.

**Comment by Rubonnek** [comment]
Wow, I've never seen anything like that before.

And by the way, thank you so much for `gdb-dashboard` -- it really makes gdb much easier to use.

**Comment by cyrus-and** [comment]
Unfortunately fetching the function bounds has always been something very hard and GDB apparently doesn't help at all. Here's the buggy code, I could revert this change (53897f8420688156a4fee4285486a6f7011d6e07) but it would break something else...

https://github.com/cyrus-and/gdb-dashboard/blob/2d31a3b391e5d0e032b791e1fb7172338b02cecb/.gdbinit#L1314-L1324

**Comment by Rubonnek** [comment]
Mmm... to find the boundaries then perhaps you could do something along the lines of:

```python
output = gdb.execute("disassemble", False, True)
function_block_ranges = re.findall("Address range (0x\w+) to (0x\w+)", output)
for entry in function_block_ranges:
    entry[0] # is the start of the block
    entry[1] # is the end of the block
```

In lieu of not having an easy to use API. Text is the universal interface after all.

**Comment by cyrus-and** [comment]
Eh, I used to do something like that (albeit without considering multiple blocks) and there were some problems too (e.g., I remember #31). But it seems that's the most reliable way to find function boundaries after all... I might consider readopting that approach.

**Comment by Rubonnek** [comment]
Yeah, it's kinda hackish and not optimal at all.

Thank you again for `gdb-dashboard` -- I wasn't aware the `gdb` module API could be such a nightmare on these cases.

**Comment by cyrus-and** [comment]
OK this is hopefully fixed now, please give it a try. :)

**Comment by Rubonnek** [comment]
Works great! Thanks!

**Comment by cyrus-and** [comment]
Awesome! Thanks for helping me reproducing this.
